### PR TITLE
FC-817 Training Provider Grouping

### PR DESCRIPTION
**Pull request by leewadhams**
- Group providers on provider search results page.
- Add 5 closest locations list to Provider details page.

### CODE REVIEW

File: src/Web/Sfa.Das.Sas.ApplicationServices/Models/GroupedProviderSearchResults.cs [hunk 1]
```diff
@@ -0,0 +1,10 @@
+﻿using System.Collections.Generic;
+
+namespace Sfa.Das.Sas.ApplicationServices.Models
+{
+    public sealed class GroupedProviderSearchResults : ProviderSearchResults
```
Comment (desderry) [inline src/Web/Sfa.Das.Sas.ApplicationServices/Models/GroupedProviderSearchResults.cs]: Should this be needed as the generic type could be used instead?

File: src/Web/Sfa.Das.Sas.ApplicationServices/Models/GroupedProviderSearchResult{T}.cs [hunk 2]
```diff
@@ -0,0 +1,15 @@
+﻿using System.Collections.Generic;
+
+namespace Sfa.Das.Sas.ApplicationServices.Models
+{
+
+    public sealed class GroupedProviderSearchResult<T>
```
Comment (desderry) [inline src/Web/Sfa.Das.Sas.ApplicationServices/Models/GroupedProviderSearchResult{T}.cs]: should this inherit from either SearchResults<t> or ProviderSearchResults<t>

File: src/Web/Sfa.Das.Sas.ApplicationServices/Models/ProviderSearchResults.cs [hunk 3]
```diff
@@ -2,7 +2,7 @@
 
 namespace Sfa.Das.Sas.ApplicationServices.Models
 {
-    public sealed class ProviderSearchResults : BaseProviderSearchResults
+    public class ProviderSearchResults : BaseProviderSearchResults
```
Comment (desderry) [inline src/Web/Sfa.Das.Sas.ApplicationServices/Models/ProviderSearchResults.cs]: think this shouldnt be plural as its as single result

File: src/Web/Sfa.Das.Sas.Shared.Components.Web/Sfa.Das.Sas.Shared.Components.Web.csproj [hunk 4]
```diff
@@ -3,6 +3,7 @@
   <PropertyGroup>
     <TargetFramework>netcoreapp2.2</TargetFramework>
     <AspNetCoreHostingModel>InProcess</AspNetCoreHostingModel>
+    <UserSecretsId>21123918-648b-445a-8f71-7ca8156967d2</UserSecretsId>
```
Comment (desderry) [inline src/Web/Sfa.Das.Sas.Shared.Components.Web/Sfa.Das.Sas.Shared.Components.Web.csproj]: whats this do? is it relevant only to your environment?

File: src/Web/Sfa.Das.Sas.Shared.Components/Mapping/TrainingProviderClosestLocationsViewModelMapper.cs [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+                Ukprn = ukprn,
+                Locations = source.Results.Hits.Select(x => new CloseLocationViewModel { Distance = x.Distance, PostCode = x.Address.Postcode, AddressWithoutPostCode = GetCommaDelimitedAddress(x.Address) }).ToList()
+            };
+
+            return model;
+        }
+
+        private string GetCommaDelimitedAddress(Address address)
+        {
+            var addressParts = new List<string>(5);
```
Comment (desderry) [inline src/Web/Sfa.Das.Sas.Shared.Components/Mapping/TrainingProviderClosestLocationsViewModelMapper.cs]: Should this be moved to the address class as used in more than one class?

File: src/Web/Sfa.Das.Sas.Shared.Components/Mapping/TrainingProviderSearchResultsItemViewModelMapper.cs [hunk 6]
```diff
@@ -17,10 +19,27 @@ public TrainingProviderSearchResultsItem Map(ProviderSearchResultItem source)
...
+                LocationAddress = GetAddress(source.Address)
             };
+
             return item;
         }
+
+        private static LocationAddress GetAddress(Address address)
+        {
+            var addressParts = new List<string>(5);
+            if (!string.IsNullOrWhiteSpace(address.Address1)) addressParts.Add(address.Address1);
```
Comment (desderry) [inline src/Web/Sfa.Das.Sas.Shared.Components/Mapping/TrainingProviderSearchResultsItemViewModelMapper.cs]: See comment above RE moving to Address class

File: src/Web/Sfa.Das.Sas.Shared.Components/Views/Shared/Components/TrainingProvider/SearchResults/Default.cshtml [hunk 7]
```diff
@@ -22,10 +22,6 @@
         <input type="hidden" asp-for="@Model.SearchQuery.Page" />
         <input type="hidden" asp-for="@Model.SearchQuery.Postcode" />
         <input type="hidden" asp-for="@Model.SearchQuery.SortOrder" />
```
Comment (desderry) [inline src/Web/Sfa.Das.Sas.Shared.Components/Views/Shared/Components/TrainingProvider/SearchResults/Default.cshtml]: Does HasNationalProviders need adding to the hidden inputs?

File: src/Web/Sfa.Das.Sas.ApplicationServices/Models/GroupedProviderSearchResult{T}.cs (hunk 2 above)
Comment (leewadhams) [reply src/Web/Sfa.Das.Sas.ApplicationServices/Models/GroupedProviderSearchResult{T}.cs]: 👍 I extracted the base class but forgot to use it here.

File: src/Web/Sfa.Das.Sas.ApplicationServices/Models/ProviderSearchResults.cs (hunk 3 above)
Comment (leewadhams) [reply src/Web/Sfa.Das.Sas.ApplicationServices/Models/ProviderSearchResults.cs]: It has a property: `IEnumerable<ProviderSearchResultItem> Hits` which has the results.

File: src/Web/Sfa.Das.Sas.Shared.Components.Web/Sfa.Das.Sas.Shared.Components.Web.csproj (hunk 4 above)
Comment (leewadhams) [reply src/Web/Sfa.Das.Sas.Shared.Components.Web/Sfa.Das.Sas.Shared.Components.Web.csproj]: Bloody VS 👿 👍

File: src/Web/Sfa.Das.Sas.Shared.Components/Mapping/TrainingProviderClosestLocationsViewModelMapper.cs (hunk 5 above)
Comment (leewadhams) [reply src/Web/Sfa.Das.Sas.Shared.Components/Mapping/TrainingProviderClosestLocationsViewModelMapper.cs]: Address is a domain model. I think this logic is very UI specific. Probably better suited in an Extension method defined in UI project.

File: src/Web/Sfa.Das.Sas.Shared.Components/Views/Shared/Components/TrainingProvider/SearchResults/Default.cshtml (hunk 7 above)
Comment (leewadhams) [reply src/Web/Sfa.Das.Sas.Shared.Components/Views/Shared/Components/TrainingProvider/SearchResults/Default.cshtml]: Don't think so. The flag for the search parameter hasn't changed, just the was we display the filter has changed.

File: src/Web/Sfa.Das.Sas.ApplicationServices/Models/GroupedProviderSearchResults.cs (hunk 1 above)
Comment (leewadhams) [reply src/Web/Sfa.Das.Sas.ApplicationServices/Models/GroupedProviderSearchResults.cs]: The base types contains properties that are returned back to the UI. This was copying the same pattern that was already there. What I have done now is redefine this class not to use the base classes as they contained numerous properties that were not relevant in this scenario.

**Pull request merged**